### ISSUE TITLE
Appsignal extension fails to load on M1 macs

### ISSUE THREAD

**Issue by grk** [appsignal/appsignal-ruby#707]
https://appsignal.com/diagnose/diag0341f8d8bcee9f83e330

```Extension installation report
  Installation result
    Status: failed
    Message: AppSignal currently does not support your system architecture (arm-darwin).Please let us know at support@appsignal.com, we aim to support everything our customers run.
  Language details
    Implementation: ruby
    Ruby version: 2.7.0-p137
  Download details
    Download URL:
    Checksum: unverified
  Build details
    Install time: 2021-02-19 08:32:11 UTC
    Architecture: arm
    Target: darwin
    Musl override: false
    Library type: static
    Source:
    Dependencies: {}
    Flags: {}
  Host details
    Root user: false
    Dependencies: {}

Host information
  Architecture: arm
  Operating System: darwin20.3.0
  Ruby version: 2.7.0-p137
  Root user: false
  Running in container:
```

Would be great to have arm-darwin support, otherwise I see this on every app load:
```
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
```

**Comment by tombruijn** [comment]
Hi @grk, we're aware of the issue and have updated our docs page about it: https://docs.appsignal.com/support/operating-systems.html#macos For now we recommend running an app through rosetta2 if you need AppSignal on an M1 mac.

This support is not on our direct roadmap. We're first going to look at ARM support for linux, as that is more likely to run in production. That too is behind some other work we're doing first. So we don't have an ETA for this at the moment.

**Comment by grk** [comment]
Would it be possible to have an ENV var or some other way to disable this error? Currently this is my console output:
```
rails parallel:prepare
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
ERROR: AppSignal failed to load extension. Please run `appsignal diagnose` and email us at support@appsignal.com
LoadError: cannot load such file -- appsignal_extension
```

**Comment by grk** [comment]
It also means that tests for code that interacts with Appsignal (like making sure an error is sent) fail locally, I need to stub out `Appsignal.send_error`

**Comment by tombruijn** [comment]
It should never result in an error. What are you testing in tests that interact with AppSignal?

**Comment by czj** [comment]
We have the same error message on all our Rails applications.

Disabling the extension (via an ENV variable or configuration) would be appreciated in dev / test environments.
Supporting the extension on Apple ARM is not important, but keeping developpers happy with less noise is :-)

**Comment by rickiesmooth** [comment]
I'm probably missing something but it seems like calling `Appsignal.increment_counter` results in an error for me.
```
2021-04-13 15:16:35 - NoMethodError - undefined method `set_string' for nil:NilClass:
15:16:35 web.1     |    /Users/ricksmit/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/appsignal-2.11.6/lib/appsignal/utils/data.rb:24:in `block in map_hash'
15:16:35 web.1     |    /Users/ricksmit/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/appsignal-2.11.6/lib/appsignal/utils/data.rb:20:in `each'
15:16:35 web.1     |    /Users/ricksmit/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/appsignal-2.11.6/lib/appsignal/utils/data.rb:20:in `map_hash'
15:16:35 web.1     |    /Users/ricksmit/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/appsignal-2.11.6/lib/appsignal/utils/data.rb:10:in `generate'
15:16:35 web.1     |    /Users/ricksmit/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/appsignal-2.11.6/lib/appsignal/helpers/metrics.rb:35:in `increment_counter'
15:16:35 web.1     |    /Users/ricksmit/dev/frontend/config/initializers/auth0_mixins_httpproxy.rb:28:in `call'
```

**Comment by tombruijn** [comment]
Hi @rickiesmooth, I've confirmed the issue. The extension Ruby class is working as expected when the extension isn't loaded, however, when generating a "Data" object it tries to do that on a `nil` object it got back from the [`method_missing` method](https://github.com/appsignal/appsignal-ruby/blob/79850976a2229ea3de7dfa662387ce3ba009b37b/lib/appsignal/extension.rb#L40). It should also check that the returned value from the stubbed Extension is valid, or check that the extension is loaded or not.

**Comment by tombruijn** [comment]
In created issue https://github.com/appsignal/appsignal-ruby/issues/719 to fix the extension error.

**Comment by tombruijn** [comment]
The extension fallback issues should be fixed in 3.0.3 cc @rickiesmooth

**Comment by rickiesmooth** [comment]
thanks @tombruijn !

**Comment by esquevin** [comment]
Hi, would it be possible to disable the extension / silence this error in dev / test environments?

It indeed creates a lot of unwanted noise. I hoped that last week changes would improve the situation but sadly my console is still littered with appsignal messages that ask me to reach your support.

Thank you, and I totally understand that supporting appsignal on M1 is not a priority.

**Comment by shairyar** [comment]
@esquevin we will take a look into that.

**Comment by tombruijn** [comment]
@esquevin and others, can you try the new "3.0.11" version of the Ruby gem. It introduces experimental Apple M1 support. Please let us know if it works here, or at support@appsignal.com with a [diagnose report](https://docs.appsignal.com/ruby/command-line/diagnose.html).

Closing the issue now. We can reopen it if new issues arise.

**Comment by esquevin** [comment]
Doesn't seem to fix the issue on my side :/ I sent a diagnose report

**Comment by tombruijn** [comment]
@esquevin For some reason your machine was recognized as ARM 64-bit Linux Musl and it tried install that version. You are running this on your Apple M1 ARM 64-bit machine, and not on that machine in a ARM 64-bit Alpine Linux Docker container?

**Comment by esquevin** [comment]
You are indeed right, I ran this from an ARM 64-bit Alpine Linux Docker container running on my M1 :/ 

My issue might not be related to the M1 then, but my dev logs are still littered with this message.

**Comment by tombruijn** [comment]
I think Alpine Linux musl is the only Linux ARM build we don't support yet. That will be a separate issue, which I've created in #742. Please subscribe to that issue for updates, but I don't know when we'll pick it up yet.

**Comment by samtgarson** [comment]
FWIW if anyone else is reading this, I was getting this with 3.0.11 but 3.0.12 seems to have fixed it for me 🤷

**Comment by tombruijn** [comment]
Ah yes, I published another fix for it in 3.0.12 after we got a report it didn't work on every Ruby install.
